### PR TITLE
[feature][callouts] Add blending mode settings

### PR DESCRIPTION
**Pull request by nirvn**
## Description

The capability to change blending mode for pretty much everything being rendered in QGIS is a great strength of ours. Let's have callouts benefit from it too:
image

### CODE REVIEW

File: src/core/callouts/qgscallout.cpp [hunk 1]
```diff
@@ -135,8 +140,24 @@ QgsCallout::DrawOrder QgsCallout::drawOrder() const
 
 void QgsCallout::render( QgsRenderContext &context, const QRectF &rect, const double angle, const QgsGeometry &anchor, QgsCalloutContext &calloutContext )
 {
-#if 0 // for debugging
   QPainter *painter = context.painter();
+  QgsScopedQPainterState painterState( painter );
+  context.setPainterFlagsUsingContext( painter );
```
Comment (nyalldawson) [inline src/core/callouts/qgscallout.cpp:145]: This isn’t required - they’ll have already been set way before we get here

File: src/core/callouts/qgscallout.cpp [hunk 2]
```diff
@@ -135,8 +140,24 @@ QgsCallout::DrawOrder QgsCallout::drawOrder() const
 
 void QgsCallout::render( QgsRenderContext &context, const QRectF &rect, const double angle, const QgsGeometry &anchor, QgsCalloutContext &calloutContext )
 {
-#if 0 // for debugging
   QPainter *painter = context.painter();
+  QgsScopedQPainterState painterState( painter );
```
Comment (nyalldawson) [inline src/core/callouts/qgscallout.cpp:144]: If you’re manually resetting the composition mode at the end of this function then you can skip this and save some cycles (or even better, make a scope painter composition mode  override class)

File: src/core/labeling/qgspallabeling.cpp [hunk 3]
```diff
@@ -563,6 +563,11 @@ void QgsPalLayerSettings::stopRender( QgsRenderContext &context )
   mRenderStarted = false;
 }
 
+bool QgsPalLayerSettings::containsAdvancedEffects() const
+{
+  return mFormat.containsAdvancedEffects() || mCallout->blendMode() != QPainter::CompositionMode_SourceOver;
```
Comment (nyalldawson) [inline src/core/labeling/qgspallabeling.cpp:568]: Would be cleaner to make a new containsAdvancedEffects virtual method in QgsCallout

**Review by nyalldawson — COMMENTED** [review]
Just some minor changes (also needs tests)

**Comment by nirvn** [comment]
@nyalldawson , thanks for the review, completely missed you did so 7 days ago :)

**Comment by nirvn** [comment]
@nyalldawson , review addressed.

File: tests/src/core/testqgscallout.cpp [hunk 4]
```diff
@@ -4109,6 +4110,22 @@ void TestQgsCallout::balloonCalloutCornerRadius()
...
+  callout->setBlendMode( QPainter::CompositionMode_Multiply );
+  QCOMPARE( callout->blendMode(), QPainter::CompositionMode_Multiply );
+  QCOMPARE( callout->containsAdvancedEffects(), true );
+
+  callout->setBlendMode( QPainter::CompositionMode_SourceOver );
+  QCOMPARE( callout->containsAdvancedEffects(), false );
+
+  callout->dataDefinedProperties().setProperty( QgsCallout::BlendMode, QStringLiteral( "multiply" ) );
+  QCOMPARE( callout->containsAdvancedEffects(), true );
+}
```
Comment (nyalldawson) [inline tests/src/core/testqgscallout.cpp:4127]: Can you add a rendering test too please?

**Comment by nirvn** [comment]
@nyalldawson , simple non-normal blend rendering test added.

**Pull request merged**